woocommerce/woocommerce
# Optimize the method that gets the downloads count for a given download

### Changes proposed in this Pull Request:

In order to retrieve the count of downloads performed for a given file in a given order there's the `WC_Customer_Download::get_download_count` method. This method was very inefficient as it was using `WC_Customer_Download_Log_Data_Store::get_download_logs_for_permission` and then locally counting the rows retrieved.

This pull request:

1. Extends `WC_Customer_Download_Log_Data_Store::get_download_count` to accept the value `count` in `$args['returns']`, this will transform the `SELECT *` query into a `SELECT COUNT` query and the method will simply return the retrieved number.
2. Adds a new `WC_Customer_Download_Log_Data_Store::get_download_logs_count_for_permission` method that uses the above change.
3. Modifies `WC_Customer_Download::get_download_count` to use the new method.



### How to test the changes in this Pull Request:

1. Create a downloadable product with one file.
2. As a customer, make an order for the product.
3. Mark the order as completed.
4. As the customer, download the file a few times.
5. Run the following in CLI (setting the proper order id) and verify that you get the proper count of downloads:

```
wp eval '$download_id=WC_Data_Store::load("customer-download")->get_downloads(["order_id"=>ORDER_ID_HERE, "return"=>"ids"])[0]; echo (new WC_Customer_Download($download_id))->get_download_count();'
```



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
